agent/hcp/telemetry_provider.go

```diff
@@ -51,28 +52,27 @@ type dynamicConfig struct {
 }
 
 // NewHCPProvider initializes and starts a HCP Telemetry provider with provided params.
-func NewHCPProvider(ctx context.Context, hcpClient client.Client, telemetryCfg *client.TelemetryConfig) (*hcpProviderImpl, error) {
-	refreshInterval := telemetryCfg.RefreshConfig.RefreshInterval
-	// refreshInterval must be greater than 0, otherwise time.Ticker panics.
-	if refreshInterval <= 0 {
-		return nil, fmt.Errorf("invalid refresh interval: %d", refreshInterval)
-	}
-
-	cfg := &dynamicConfig{
-		Endpoint:        telemetryCfg.MetricsConfig.Endpoint,
-		Labels:          telemetryCfg.MetricsConfig.Labels,
-		Filters:         telemetryCfg.MetricsConfig.Filters,
-		RefreshInterval: refreshInterval,
-	}
-
+func NewHCPProvider(ctx context.Context, hcpClient client.Client) *hcpProviderImpl {
 	t := &hcpProviderImpl{
-		cfg:       cfg,
+		// Initialize with default config values.
+		cfg: &dynamicConfig{
+			Labels:          map[string]string{},
+			Filters:         client.DefaultMetricFilters,
+			RefreshInterval: defaultTelemetryConfigRefreshInterval,
+			Endpoint:        nil,
+			Enabled:         false,
+		},
 		hcpClient: hcpClient,
 	}
 
-	go t.run(ctx, refreshInterval)
+	// Try to initialize the config once before running periodic fetcher.
+	if newCfg := t.getUpdate(ctx); newCfg != nil {
+		t.cfg = newCfg
```

Comment:
I think this is redundant since it's set in `getUpdate`:

https://github.com/hashicorp/consul/blob/9ed67e439632f3e29f0b8c1fcf548be7f3588b9d/agent/hcp/telemetry_provider.go#L129

relatedly, I think `getUpdate` could be renamed to something like `updateConfig` since it's mutating config internally

Reply:
100%, I also renamed the method, thanks for catching this